tests/DCPS/XTypes/XTypes.cpp

```diff
@@ -0,0 +1,668 @@
+    get_topic(ts, dp, topic_name, topic, registered_type_name);
+  } else if (type == "ModifiedTypeMutableUnion") {
+    ModifiedTypeMutableUnionTypeSupport_var ts = new ModifiedTypeMutableUnionTypeSupportImpl;
+    get_topic(ts, dp, topic_name, topic, registered_type_name);
+  } else if (type == "ModifiedNameMutableUnion") {
+    ModifiedNameMutableUnionTypeSupport_var ts = new ModifiedNameMutableUnionTypeSupportImpl;
+    get_topic(ts, dp, topic_name, topic, registered_type_name);
+  } else if (type == "Trim64Struct") {
+    Trim64StructTypeSupport_var ts = new Trim64StructTypeSupportImpl;
+    get_topic(ts, dp, topic_name, topic, registered_type_name);
+  } else if (type == "Trim20Struct") {
+    Trim20StructTypeSupport_var ts = new Trim20StructTypeSupportImpl;
+    get_topic(ts, dp, topic_name, topic, registered_type_name);
+  } else if (type == "AppendableStructWithDependency") {
+    AppendableStructWithDependencyTypeSupport_var ts = new AppendableStructWithDependencyTypeSupportImpl;
+    get_topic(ts, dp, topic_name, topic, registered_type_name);
+  } else {
+    ACE_ERROR((LM_ERROR, "ERROR: Type %s is not supported\n", type.c_str()));
+    return 1;
+  }
+
+  if (writer) {
+    Publisher_var pub = dp->create_publisher(PUBLISHER_QOS_DEFAULT, 0,
+                                             DEFAULT_STATUS_MASK);
+    DataWriterQos dw_qos;
+    pub->get_default_datawriter_qos(dw_qos);
+    dw_qos.durability.kind = TRANSIENT_LOCAL_DURABILITY_QOS;
+
+    DataWriter_var dw = pub->create_datawriter(topic, dw_qos, 0,
+      DEFAULT_STATUS_MASK);
+
+    ACE_OS::sleep(ACE_Time_Value(1, 0));
+
+    failed = !check_inconsistent_topic_status(topic);
+
+    if (!failed && !expect_to_fail) {
+      if (type == "PlainCdrStruct") {
+        write_plain_cdr_struct(dw);
+      } else if (type == "FinalStruct") {
+        write_final_struct(dw);
+      } else if (type == "ModifiedFinalStruct") {
+        write_modified_final_struct(dw);
+      } else if (type == "AppendableStruct") {
+        write_appendable_struct(dw);
+      } else if (type == "AdditionalPrefixFieldStruct") {
+        write_additional_prefix_field_struct(dw);
+      } else if (type == "AdditionalPostfixFieldStruct") {
+        write_additional_postfix_field_struct(dw);
+      } else if (type == "MutableStruct") {
+        write_mutable_struct(dw);
+      } else if (type == "ModifiedMutableStruct") {
+        write_modified_mutable_struct(dw);
+      } else if (type == "MutableUnion") {
+        write_mutable_union(dw);
+      } else if (type == "ModifiedMutableUnion") {
+        write_modified_mutable_union(dw);
+      } else if (type == "Trim64Struct") {
+        write_trim64_struct(dw);
+      } else if (type == "AppendableStructWithDependency") {
+        write_appendable_struct_with_dependency(dw);
+      }
+    }
+
+    ACE_OS::sleep(ACE_Time_Value(3, 0));
+
+    if (failed) {
+      ACE_ERROR((LM_ERROR, "ERROR: Writer failed for type %s\n", type.c_str()));
+    }
```

Comment:
I think this hasn't handled all possible cases for `failed` and `expect_to_fail`. It looks to me that the test fails when `failed != expect_to_fail`. Also `failed == true` doesn't always indicate a failed test as implied by the if block at lines 607-609, that is if `failed` is `true` and `expect_to_fail` is also `true`, then the test should be considered passed. 

So I think we could do something like, if `failed != expect_to_fail` then the test failed. Otherwise, if `failed == expect_to_fail` and both are `false`, then perform write/read message depending on whether it is a writer or reader; if `failed == expect_to_fail` but both are `true`, then the test passed for this negative test case and does nothing else.

Same comment applies to the similar logic for the reader case.

Reply:
Part of this logic is handled within `check_inconsistent_topic_status` function, but I've made some changes, and also replaced `expect_to_fail` flag name with 'expect_to_match', as Adam suggested earlier, for clarity.